pkg/controller/restore_controller.go

```diff
@@ -506,48 +506,74 @@ func (c *restoreController) fetchFromBackupStorage(backupName string, pluginMana
-		if err != nil {
-			logContext.WithError(errors.WithStack(err)).Error("Unable to get an object store in the location ", location.Name)
-			continue
+		info, err = c.backupInfoForLocation(location, backupName, pluginManager)
+		if err == nil {
+			break
 		}
+		log := logContext.WithField("locationName", location.Name)
+		log.WithError(errors.WithStack(err)).Error("Unable to fetch a backup from object storage")
+	}
 
-		info.backup, err = c.getBackup(info.objectStore, location.Spec.ObjectStorage.Bucket, backupName)
-		if err != nil {
-			logContext.WithError(errors.WithStack(err)).Error("Unable to find a backup in the location ", location.Name)
-			continue
-		}
+	if info == (backupInfo{}) {
+		return backupInfo{}, errors.New("not able to fetch from backup storage")
+	}
 
-		// ResourceVersion needs to be cleared in order to create the object in the API
-		info.backup.ResourceVersion = ""
-		// Clear out the namespace, in case the backup was made in a different cluster, with a different namespace
-		info.backup.Namespace = ""
+	return info, nil
+}
 
-		info.backup, err = c.backupClient.Backups(c.namespace).Create(info.backup)
-		if err != nil {
-			logContext.WithError(errors.WithStack(err)).Error("Unable to create API object for Backup")
-		} else {
-			// backup found in a location's object store and successfully created in the API
-			info.bucketName = location.Spec.ObjectStorage.Bucket
-			return info, nil
+func orderedBackupLocations(locations []*api.BackupStorageLocation, defaultLocationName string) []*api.BackupStorageLocation {
+	var result []*api.BackupStorageLocation
+
+	for i := range locations {
+		if locations[i].Name == defaultLocationName {
+			// put the default location first
+			result = append(result, locations[i])
+			// append everything before the default
+			result = append(result, locations[:i]...)
+			// append everything after the default
+			result = append(result, locations[i+1:]...)
+
+			return result
 		}
 	}
 
-	return backupInfo{}, errors.New("not able to fetch from backup storage")
+	return locations
+}
+
+func (c *restoreController) backupInfoForLocation(location *api.BackupStorageLocation, backupName string, pluginManager plugin.Manager) (backupInfo, error) {
+	objectStore, err := getObjectStoreForLocation(location, pluginManager)
+	if err != nil {
+		return backupInfo{}, err
+	}
+
+	backup, err := c.getBackup(objectStore, location.Spec.ObjectStorage.Bucket, backupName)
+	if err != nil {
+		return backupInfo{}, err
+	}
+
+	// ResourceVersion needs to be cleared in order to create the object in the API
+	backup.ResourceVersion = ""
+	// Clear out the namespace, in case the backup was made in a different cluster, with a different namespace
+	backup.Namespace = ""
+
+	backupCreated, _ := c.backupClient.Backups(c.namespace).Create(backup)
+	if err == nil {
+		backup = backupCreated
+	}
+	bucketName := location.Spec.ObjectStorage.Bucket
+
+	info := backupInfo{
```

Comment:
I think putting together this struct in parts all over the inside of this block was a bad idea. I like it better using local variables to fetch the values and then doing a single construction here. It feels a lot more explicit and reads better to me. And even so the bucket name is "right there", I think being consistent with all the variables/elements trumps aiming to have less code (which I also like).

Reply:
I think I generally agree. Though I do prefer that if we're doing something like the following:

```
bucketName := location.Spec.ObjectStorage.Bucket

info := backupInfo{
    bucketName: bucketName,
    ....
}
```

that we skip the declaration of `bucketName` and just do:
```
info := backupInfo{
    bucketName: location.Spec.ObjectStorage.Bucket,
    ....
}
```

Reply:
One less variable to keep track of :)